src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,259 @@
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Replica assignment has brokers that are not alive",
+      [&all_node_ids](const reassignable_partition& partition) {
+          auto unkown_broker_id_it = std::find_if(
+            partition.replicas->begin(),
+            partition.replicas->end(),
+            [all_node_ids](const model::node_id& node_id) {
+                return std::find(
+                         all_node_ids.begin(), all_node_ids.end(), node_id)
+                       == all_node_ids.end();
+            });
+          return unkown_broker_id_it == partition.replicas->end();
+      });
+
+    // Store any invalid partitions in the response
+    topic_response.partitions = std::move(invalid_partitions);
+    resp_it = std::move(topic_response);
+
+    return valid_partitions_end;
+}
+
+template<>
+ss::future<response_ptr> alter_partition_reassignments_handler::handle(
+  request_context ctx, [[maybe_unused]] ss::smp_service_group ssg) {
+    alter_partition_reassignments_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    log_request(ctx.header(), request);
+    alter_partition_reassignments_response resp;
+
+    if (!ctx.authorized(
+          security::acl_operation::alter, security::default_cluster_name)) {
+        resp.data.error_code = error_code::cluster_authorization_failed;
+        resp.data.error_message = ss::sstring{
+          error_code_to_str(error_code::cluster_authorization_failed)};
+        co_return co_await ctx.respond(std::move(resp));
+    }
+
+    // TODO(@NyaliaLui) - Check if we are the controller
+
+    resp.data.responses.reserve(request.data.topics.size());
+    auto all_node_ids = ctx.metadata_cache().node_ids();
+
+    for (reassignable_topic& topic : request.data.topics) {
+        reassignable_topic_response topic_response{.name = topic.name};
+
+        auto valid_partitions_end = validate_partitions(
+          topic.partitions.begin(),
+          topic.partitions.end(),
+          std::back_inserter(resp.data.responses),
+          topic_response,
+          all_node_ids);
+
+        for (auto it = topic.partitions.begin(); it != valid_partitions_end;
+             ++it) {
+            const reassignable_partition& partition{*it};
+            model::ntp ntp{
+              model::kafka_namespace, topic.name, partition.partition_index};
+
+            if (partition.replicas.has_value()) {
+                auto assignments
+                  = co_await ctx.topics_frontend().generate_reassignments(
+                    partition.partition_index, topic.name, *partition.replicas);
+                if (
+                  assignments.has_value()
+                  && assignments.value().front().id
+                       == partition.partition_index) {
```

Comment:
can you write a comment here about what this case is? i can't seem to figure it out after thinking about it for a while. why would the first value be significant? also, should there be a check that assignments is non-empty before grabbing its first element?